contracts/protocol/v1/traders/P1InverseOrders.sol

```diff
@@ -0,0 +1,495 @@
+    mapping (bytes32 => OrderStatus) public _STATUS_;
+
+    // ============ Constructor ============
+
+    constructor (
+        address perpetualV1,
+        uint256 chainId
+    )
+        public
+    {
+        _PERPETUAL_V1_ = perpetualV1;
+
+        /* solium-disable-next-line indentation */
+        _EIP712_DOMAIN_HASH_ = keccak256(abi.encode(
+            EIP712_DOMAIN_SEPARATOR_SCHEMA_HASH,
+            keccak256(bytes(EIP712_DOMAIN_NAME)),
+            keccak256(bytes(EIP712_DOMAIN_VERSION)),
+            chainId,
+            address(this)
+        ));
+    }
+
+    // ============ External Functions ============
+
+    /**
+     * @notice Allows an account to take an order cryptographically signed by a different account.
+     * @dev Emits the LogOrderFilled event.
+     *
+     * @param  sender  The address that called the trade() function on PerpetualV1.
+     * @param  maker   The maker of the order.
+     * @param  taker   The taker of the order.
+     * @param  price   The current oracle price of the underlying asset.
+     * @param  data    A struct of type TradeData.
+     * @return         The assets to be traded and traderFlags that indicate that a trade occurred.
+     */
+    function trade(
+        address sender,
+        address maker,
+        address taker,
+        uint256 price,
+        bytes calldata data,
+        bytes32 /* traderFlags */
+    )
+        external
+        returns (P1Types.TradeResult memory)
+    {
+        address perpetual = _PERPETUAL_V1_;
+
+        require(
+            msg.sender == perpetual,
+            "msg.sender must be PerpetualV1"
+        );
+
+        if (taker != sender) {
+            require(
+                P1Getters(perpetual).hasAccountPermissions(taker, sender),
+                "Sender does not have permissions for the taker"
+            );
+        }
+
+        TradeData memory tradeData = abi.decode(data, (TradeData));
+        bytes32 orderHash = _getOrderHash(tradeData.order);
+
+        // validations
+        _verifyOrderStateAndSignature(
+            tradeData,
+            orderHash
+        );
+        _verifyOrderRequest(
+            tradeData,
+            maker,
+            taker,
+            perpetual,
+            price
+        );
+
+        // set _FILLED_AMOUNT_
+        uint256 oldFilledAmount = _FILLED_AMOUNT_[orderHash];
+        uint256 newFilledAmount = oldFilledAmount.add(tradeData.fill.amount);
+        require(
+            newFilledAmount <= tradeData.order.amount,
+            "Cannot overfill order"
+        );
+        _FILLED_AMOUNT_[orderHash] = newFilledAmount;
+
+        // Inverse perpetual: The fill amount is denoted in quote currency (position).
+        emit LogOrderFilled(
+            orderHash,
+            tradeData.order.flags,
+            tradeData.order.triggerPrice,
+            tradeData.fill
+        );
+
+        // `isBuyOrder` is from the maker's perspective, and relative to the base currency.
```

Comment:
Do you think `, and relative to the base currency.` is clear enough? If not, you can revert this line to the old version

```
 // `isBuyOrder` is from the maker's perspective.
```